Create new Code Origins doc page

sstonehill12 opened the pull request:
<!-- *Note: Please remember to review the Datadog Documentation [Contribution Guidelines](https://github.com/DataDog/documentation/blob/master/CONTRIBUTING.md) if you have not yet done so.* -->

### What does this PR do? What is the motivation?
<!-- A brief description of the change being made with this pull request. What is your motivation for the PR? -->
Add new public docs page for Code Origins feature, and associated images.

### Merge instructions

<!-- 
If you're waiting for a release or there are other considerations that you want us to be aware of, list them here. 
If the PR is ready to be merged once it receives the required reviews, check the box below after you've created the PR.
-->

Merge readiness:
- [x] Ready for merge

Merge queue is enabled in this repo. To have it automatically merged after it receives the required reviews, create the PR (from a branch that follows the `<yourname>/description` naming convention) and then add the following PR comment:

```
/merge
```

### Additional notes
<!-- Anything else we should know when reviewing?-->

<!-- Previewing the PR: Assuming you are a Datadog employee and named your branch `<yourname>/<description>`, a preview build will run and links to the preview output will be auto-generated and posted in the PR comments. The links will 404 until the preview build is finished running. -->

evanchooly commented on content/en/tracing/guide/code_origins.md:
```diff
@@ -0,0 +1,159 @@
+import datadog.trace.api.DDTrace;
+
+// Enable/disable Code Origins
+DDTrace.enableCodeOrigins();
```
Where did this come from?  I can't find any trace (pun intended) of this type in `dd-trace-java`.

sstonehill12 replied:
Yep this was definitely made up by Cursor! I'll swap it out with the JVM flag instructions.

evanchooly commented on content/en/tracing/guide/code_origins.md:
````diff
@@ -0,0 +1,159 @@
+
+```bash
+# Enable/disable Code Origins
+DD_TRACE_CODE_ORIGINS_ENABLED=true
````
at least in the java tracer, this was renamed to DD_CODE_ORIGIN_FOR_SPANS_ENABLED.  Alternately, the flag `-Ddd.code.origin.for.spans.enabled=true` can be passed to the JVM at start up.

sstonehill12 replied:
Will update!

sstonehill12 replied:
Decided to remove the language-specific instructions and keep the environment variable instructions only - for simplicity!

tylfin commented on content/en/tracing/guide/code_origins.md:
```diff
@@ -0,0 +1,159 @@
+
+## Overview
+
+Code Origins is a feature that helps you understand where your APM spans originate from in your codebase. When enabled, it automatically captures the file path, line number, and function name where each span was created, giving you deeper visibility into your application's execution flow.
+
+This information is particularly valuable when:
+- Debugging performance issues
+- Understanding which parts of your code are being executed
+- Identifying hot spots in your application
+- Onboarding new team members to your codebase
```
```suggestion
Code Origins captures the exact location in your codebase where APM spans are created. When enabled, it automatically adds file path, line number, and function name to each span, making it easier to:

- Debug performance issues
- Understand code execution flow
- Identify performance bottlenecks
- Onboard new team members
```

tylfin commented on content/en/tracing/guide/code_origins.md:
```diff
@@ -0,0 +1,159 @@
+## Getting started
+
+### Prerequisites
+- Set up [Datadog APM][6] to capture spans and view them in the Traces Explorer.
+- Service must meet the [Compatibility Requirements](#compatibility-requirements) below.
+- (Recommended) For the best experience, enable [Source Code Integration][7] to see Code Origin application code previews.
```
```suggestion
- [Datadog APM][6] configured to capture spans
- Service meets the [Compatibility Requirements](#compatibility-requirements) below
- [Source Code Integration][7] enabled (recommended for code previews)
```

tylfin commented on content/en/tracing/guide/code_origins.md:
```diff
@@ -0,0 +1,159 @@
+
+### Enable Code Origins
+
+You can enable Code Origins through environment variables or Datadog tracing library configuration:
```
```suggestion
Enable Code Origins through environment variables or tracing library configuration:
```

tylfin commented on content/en/tracing/guide/code_origins.md:
```diff
@@ -0,0 +1,159 @@
+
+## How it works
+
+**Entry spans**: The Code Origin is the first method in the application code that handles the incoming request. When a service is [instrumented][6], we use the APM integration metadata to identify the source code location where incoming requests enter the system.
+
+**Exit spans**: The Code Origin is the specific line of code that made the outgoing request to a downstream service/component. By examining the call stack, we can identify the line of code where an exit span starts, instrument that line, and decorate the span accordingly. Third-party code is skipped when walking the call stack. 
```
```suggestion
Code Origins captures two types of spans:

**Entry spans**: Shows the first method in your application code that handles an incoming request. The APM integration identifies the source code location where requests enter your system.

**Exit spans**: Shows the exact line of code that makes an outgoing request to a downstream service. The tracer examines the call stack to identify the line where an exit span starts, skipping third-party code.
```

tylfin commented:
I generated some language suggestions feeding your changes with the RFC and CONTRIBUTING guide through an LLM. I thought these were generally good, but feel free to disregard them.

Also, the build preview is giving a 4xx because the CI is failing, but locally it runs just fine. I would try pulling the latest changes from the main branch and see if that helps 🤷

tylfin commented on content/en/tracing/guide/code_origins.md:
````diff
@@ -0,0 +1,117 @@
+Run your service with the below environment variable:
+
+```bash
+# Enable/disable Code Origins
+DD_CODE_ORIGIN_FOR_SPANS_ENABLED=true
````
```suggestion
export DD_CODE_ORIGIN_FOR_SPANS_ENABLED=true
```

tylfin commented on content/en/tracing/guide/code_origins.md:
````diff
@@ -0,0 +1,117 @@
+
+Run your service with the below environment variable:
+
+```bash
````
I don't know why, but if you make it `shell` instead of `bash`, it will render with a Copy button

```suggestion
```shell
```

evanchooly commented on content/en/tracing/guide/code_origins.md:
```diff
@@ -0,0 +1,117 @@
+---
+title: Code Origins for Spans
+description: "Learn how to use Code Origins to understand where your spans originate from in your codebase"
```
"from" is unnecessary here

tylfin replied:
```suggestion
description: "Learn how to use Code Origins to understand where your spans originate in your codebase"
```

evanchooly commented on content/en/tracing/guide/_index.md:
```diff
@@ -82,4 +82,5 @@ These tutorials walk you through setting up a sample multi-service application w
     {{< nextlink href="/tracing/guide/setting_up_apm_with_kubernetes_service/" >}}Setting up APM with Kubernetes Service{{< /nextlink >}}
     {{< nextlink href="/tracing/guide/aws_payload_tagging/" >}}Capture requests and responses from AWS services automatically{{< /nextlink >}}
     {{< nextlink href="/tracing/guide/service_overrides/" >}}Service Overrides{{< /nextlink >}}
+    {{< nextlink href="/tracing/guide/code_origins/" >}}Capture Code Origins for spans{{< /nextlink >}}
```
I'm not sure about the "Capture" here.  Makes more sense, to me, to just have "Code Origins for spans."  And maybe capitalize "spans" here if this is supposed to be a section title.

tylfin replied:
```suggestion
    {{< nextlink href="/tracing/guide/code_origins/" >}}Code Origins for Spans{{< /nextlink >}}
```

cswatt commented:
Added DOCS-10601 to track Documentation team review.

tylfin commented:
cc @cswatt this is ready for review

sstonehill12 commented on content/en/tracing/guide/code_origins.md:
```diff
@@ -0,0 +1,124 @@
+|---|---|---|
+| Java | `1.47.0` or later | Spring Boot/Data, gRPC servers, Micronaut 4, Kafka consumers|
+| Python | `2.15.0` or later | Django, Flask, Starlette and derivatives|
+| Node.js | `4.49.0` or later | Fastify (Express is coming soon!) |
```
@watson I updated the compatibility requirements for Node.js to exclude support for exit spans for now. Can you confirm if this looks correct? (incl. tracer version)

domalessi commented on content/en/tracing/guide/_index.md:
```diff
@@ -82,4 +82,5 @@ These tutorials walk you through setting up a sample multi-service application w
     {{< nextlink href="/tracing/guide/setting_up_apm_with_kubernetes_service/" >}}Setting up APM with Kubernetes Service{{< /nextlink >}}
     {{< nextlink href="/tracing/guide/aws_payload_tagging/" >}}Capture requests and responses from AWS services automatically{{< /nextlink >}}
     {{< nextlink href="/tracing/guide/service_overrides/" >}}Service Overrides{{< /nextlink >}}
+    {{< nextlink href="/tracing/guide/code_origins/" >}}Code Origins for Spans{{< /nextlink >}}
```
Also... Just want to confirm that the intention is to add the page to Guides and not include it in the left-hand nav? Guides are typically about more detailed use cases as opposed to describing how to use a sub-product/feature. I think Guides is fine if you're cool with the content being a bit more hidden.

sstonehill12 replied:
I just pushed a change to add it to the left-hand nav menu as well! Thanks for the reminder.

sstonehill12 commented on content/en/tracing/guide/code_origins.md:
```diff
@@ -0,0 +1,127 @@
+
+### Enable Code Origins
+
+Run your service with the below environment variable:
```
For this one, I think we should keep it closer to the original instruction. The key is that they are including that env variable when they run their service, but depending on the language and configuration setup, it might not be run as a command per se.

domalessi replied:
Totally fair. In that case, let's just switch to "the following environment variable"? We try to avoid directional terms in the docs!

sstonehill12 commented on content/en/tracing/guide/code_origins.md:
```diff
@@ -0,0 +1,127 @@
+
+## Using Code Origins
+
+### In the Trace Explorer
+
+1. Navigate to the [Trace Explorer][1].
```
Was there meant to be a suggested code change here?

domalessi commented on static/images/tracing/guide/code_origins/code_origin_add_span_tags_spotlight.png:
Can we also put a little spotlight box around Code Origin?

domalessi commented on content/en/tracing/guide/code_origins.md:
```diff
@@ -0,0 +1,127 @@
+1. Verify Code Origins is [enabled](#enable-code-origins) in your tracing library configuration.
+1. Confirm that your service meets all [compatibility requirements](#compatibility-requirements).
+    - In particular, check whether your service's language and framework support Code Origins for entry and/or exit spans.
+1. Filter for spans that include Code Origins using the query `@_dd.code_origin.type:* in the [Trace Explorer][1].
```
```suggestion
1. Filter for spans that include Code Origins using the query `@_dd.code_origin.type:*` in the [Trace Explorer][1].
```

domalessi commented on content/en/tracing/guide/code_origins.md:
```diff
@@ -0,0 +1,127 @@
+
+Code Origins identifies the source code locations where APM spans originate. Code Origin details are captured for two types of spans:
+
+**Entry spans**: The first method in your application code that handles an incoming request. The APM integration pinpoints the line of code where requests enter your system.
```
Does this make sense? Trying to align "entry span(s)" and "first method(s)" in terms of singular/plural. 

```suggestion
**Entry spans:** These represent the first method(s) in your application code that handle incoming requests. The APM integration pinpoints the line of code where each request enters your system.
```

domalessi commented on content/en/tracing/guide/code_origins.md:
```diff
@@ -0,0 +1,127 @@
+
+**Entry spans**: The first method in your application code that handles an incoming request. The APM integration pinpoints the line of code where requests enter your system.
+
+**Exit spans**: The exact line of code that makes an outgoing request to a downstream service. The tracer examines the call stack to identify the line where an exit span starts, skipping third-party code.
```
Same thing here. Trying to align singular/plural. Does this make sense? 

```suggestion
**Exit spans:** These represent the lines of code in your application that make outgoing requests to downstream services. The tracer examines the call stack to identify where each exit span originates, skipping third-party code.
```

domalessi reviewed (commented):
Just a few minor things!

evanchooly commented on content/en/tracing/guide/code_origins.md:
```diff
@@ -89,9 +89,9 @@ export DD_CODE_ORIGIN_FOR_SPANS_ENABLED=true
+**Entry spans:** These represent the first method(s) in your application code that handle incoming requests. The APM integration pinpoints the line of code where each request enters your system.
 
-**Exit spans**: The exact line of code that makes an outgoing request to a downstream service. The tracer examines the call stack to identify the line where an exit span starts, skipping third-party code.
+**Exit spans:** These represent the lines of code in your application that make outgoing requests to downstream services. The tracer examines the call stack to identify where each exit span originates, skipping third-party code.
```
mentioning the stack walking is an implementation detail that's unnecessary and I'm not sure it's 100% applicable across all tracers.

sstonehill12 replied:
@evanchooly This is true, but we do have a note below as a disclaimer that some tracing libraries may have slightly different implementations. The goal for this "How it works" section is to give a little more context for curious users without information overload. I'm open to wording changes if you have suggestions!

evanchooly commented on content/en/tracing/guide/code_origins.md:
```diff
@@ -89,9 +89,9 @@ export DD_CODE_ORIGIN_FOR_SPANS_ENABLED=true
 Code Origins identifies the source code locations where APM spans originate. Code Origin details are captured for two types of spans:
 
-**Entry spans**: The first method in your application code that handles an incoming request. The APM integration pinpoints the line of code where requests enter your system.
+**Entry spans:** These represent the first method(s) in your application code that handle incoming requests. The APM integration pinpoints the line of code where each request enters your system.
```
These two sentences are basically saying the same thing but in different words.  They're also a bit repetitive.

sstonehill12 replied:
@evanchooly I don't disagree! But in an effort to explain "how it works" at a high-level, is there a better way you can think of to describe this?

sstonehill12 replied:
Customer questions about how our products work are really common actually - which is why I actually wanted to include this section.

From my limited experience so far, understanding the "how it works" is often a prerequisite for customers enabling the feature at all.

I'd like to have this section in the docs that we can at least link to, and we can improve the content as we get feedback from customers.

duviduduvid commented on content/en/tracing/guide/code_origins.md:
```diff
@@ -0,0 +1,127 @@
+
+1. Navigate to the [Trace Explorer][1].
+2. Click on any trace to view its details.
+3. In the span details panel, look for the "Code Origin" section.
```
(nit picky) Shouldn't it say "...look for the "Code Origin" section in the **Overview** tab" ?

sstonehill12 replied:
I didn't include that detail on purpose, because there is some discussion with the APM traces team about whether it will be moved to another tab!

duviduduvid commented on content/en/tracing/guide/code_origins.md:
```diff
@@ -0,0 +1,127 @@
+
+If Code Origin information is missing:
+
+1. Verify Code Origins is [enabled](#enable-code-origins) in your tracing library configuration.
+1. Confirm that your service meets all [compatibility requirements](#compatibility-requirements).
+    - In particular, check whether your service's language and framework support Code Origins for entry and/or exit spans.
+1. Filter for spans that include Code Origins using the query `@_dd.code_origin.type:*` in the [Trace Explorer][1].
+1. Enable [Source Code Integration][7] to see code previews in the APM Trace details.
```
Why all of these numbered as 1 ?

OmerRaviv replied:
```suggestion
1. Enable [Source Code Integration][7] to see code previews in the APM Trace side panel.
```

duviduduvid commented on static/images/tracing/guide/code_origins/code_origin_details_spotlight.png:
Perhaps worth retaking this screenshot because the design of the code section has changed a bit - its background is grey instead of white.

sstonehill12 replied:
Ah it's such a minor difference, especially with the spotlight effect.. I don't think customers will notice the difference to be honest! I'll wait to update the image when there's a more significant UI change :)

duviduduvid reviewed (approved):
👍🏽 Left only minor comments

OmerRaviv commented on content/en/tracing/guide/code_origins.md:
```diff
@@ -0,0 +1,127 @@
+| Java | 1.47.0+ | Spring Boot/Data, gRPC servers, Micronaut 4, Kafka consumers (entry spans only)|
+| Python | 2.15.0+ | Django, Flask, Starlette and derivatives (entry and exit spans)|
+| Node.js | 4.49.0+ | Fastify (entry and exit spans), all other [APM-supported integrations][8] (exit spans only)|
+| .NET | 3.15.0+ | ASP.NET, ASP.NET Core (entry and exit spans)| 
```
```suggestion
| .NET | 3.15.0+ | ASP.NET, ASP.NET Core (entry spans only)| 
```

OmerRaviv replied:
See release notes for reference https://github.com/DataDog/dd-trace-dotnet/releases

sstonehill12 replied:
@OmerRaviv I got confirmation from Matan last week that .NET supports both entry and exit! [Slack thread](https://dd.slack.com/archives/G01AQ0HN9C3/p1744750280167169)

sstonehill12 replied:
Just spoke with @OmerRaviv and we believe exit spans are disabled for .NET at this time. Will confirm with Matan offline!

OmerRaviv commented on content/en/tracing/guide/code_origins.md:
```diff
@@ -0,0 +1,127 @@
+  
+    {{< img src="tracing/guide/code_origins/code_origin_add_span_tag_spotlight.png" alt="Code Origin - Add span tag with Dynamic Instrumentation" style="width:100%;">}}  
+
+### In your IDE
```
We should specify which IDEs here - I believe right now this only works in IntelliJ? Maybe also WebStorm?

sstonehill12 replied:
@OmerRaviv I actually purposefully removed the specific IDE names so that the IDE Integration pages can be the source of truth for Code Origins support! Wanted to minimize risk of having stale info on our page. But I can add it back if you feel strongly. LMK, otherwise I'll keep as is.

OmerRaviv commented on content/en/tracing/guide/code_origins.md:
```diff
@@ -0,0 +1,127 @@
+
+**Exit spans:** These represent the lines of code in your application that make outgoing requests to downstream services. The tracer examines the call stack to identify where each exit span originates, skipping third-party code.
+
+<div class="alert alert-info">Note: Some tracing libraries may have slightly different implementations to optimize for performance.</div>
```
```suggestion
```
This seems too vague to be useful - and also might flare up performance concerns from customers needlessly. Instead, maybe we should just say that exit spans are only supported in some runtimes?

sstonehill12 replied:
@OmerRaviv This note wasn't actually meant to refer to entry/exit span support! It was more of a disclaimer for any tracers that don't do the stack walk method to identify exit spans. 

I can definitely call out the specific tracer libraries that do stack walk vs. another method - I just need to know which ones!

My goal is to hit a balance of an acceptable "How it works" explanation without information overload. LMK what you think!

sstonehill12 replied:
Given decision to exclude mention of exit spans in the docs for now, I'll remove this How it works section as well, since there's not enough additional info it provides without the Exit spans definition.

OmerRaviv commented on content/en/tracing/guide/code_origins.md:
```diff
@@ -0,0 +1,127 @@
+
+**Performance Impact**: Code Origins adds negligible performance overhead in production environments.
+
+**Pricing Impact**: While in Preview, Code Origins has no impact on billing. Any changes related to Code Origins that have an impact on pricing will be communicated before going into effect.
```
This seems like to much of a deterrent - can we make a decision now about whether we want to charge for it or no now, and then either alter or delete this text? My concern is that we will scare aware customers now, for the vague concern that we *might* want to charge for it later.
```suggestion.
```

sstonehill12 replied:
@OmerRaviv From on my conversation with Antoine, he is against making this a forever-free feature, given that the extra CO metadata could double or triple the # bytes for a given span ([example](https://dd.datad0g.com/api/ui/trace/67f7c0e500000000a447c60608e74052)) and DD has to eat the cost if we don't charge for that extra data.

I'd argue that customers would be more hesitant to enable it if we say nothing about how it will impact their billing.

I could revise the second sentence to give a bit more comfort around the ability to opt-out once it goes into GA.

sstonehill12 replied:
Spoke with @OmerRaviv and we're going to remove this section for now until we have more final answers about exit spans/billing implications.

OmerRaviv commented on content/en/tracing/guide/code_origins.md:
```diff
@@ -0,0 +1,127 @@
+
+<div class="alert alert-info">Note: Some tracing libraries may have slightly different implementations to optimize for performance.</div>
+
+## Impact on performance and billing
```
Does this really deserve its own section? Maybe fold into 'How it works'?

sstonehill12 replied:
Decided to remove this section altogether for now.

OmerRaviv commented on content/en/tracing/guide/code_origins.md:
```diff
@@ -0,0 +1,127 @@
+1. Filter for spans that include Code Origins using the query `@_dd.code_origin.type:*` in the [Trace Explorer][1].
+1. Enable [Source Code Integration][7] to see code previews in the APM Trace details.
+
+**Note**: Code Origins does not capture third-party code.
```
```suggestion
```

Potentially too confusing.

sstonehill12 replied:
Agree!

OmerRaviv reviewed (approved):
LGTM, left a few comments

sstonehill12 commented on content/en/tracing/guide/code_origins.md:
```diff
@@ -0,0 +1,127 @@
+
+**Performance Impact**: Code Origins adds negligible performance overhead in production environments.
+
+**Pricing Impact**: While in Preview, Code Origins has no impact on billing. Any changes related to Code Origins that have an impact on pricing will be communicated before going into effect.
```
Removed this section in latest version.

sstonehill12 commented:
@domalessi Ready for a final look-through! Biggest changes to note: 
- Removed all mention of support for Exit Spans (we are still evaluating whether we will support this feature long-term).
- Removed the "How it works" and "Impact on performance and pricing" sections (will add back after decision on exit spans).

sstonehill12 commented on content/en/tracing/guide/code_origins.md:
```diff
@@ -0,0 +1,112 @@
+
+To submit questions, feedback, or requests related to Code Origins, fill out <a href="https://docs.google.com/forms/d/e/1FAIpQLScyeRsF2GJjYdf9bUyeDjt8_9id-gvqiBU1SHR3ioDGe5eF3g/viewform?usp=header">this form</a> with details. 
+
+For time-sensitive issues, contact <a href="https://www.datadoghq.com/support/">Datadog support</a>.
```
```suggestion
```

sstonehill12 commented on content/en/tracing/guide/code_origins.md:
```diff
@@ -0,0 +1,112 @@
+Code Origins is currently in Preview. To join the preview, follow the instructions below. 
+
+To submit questions, feedback, or requests related to Code Origins, fill out <a href="https://docs.google.com/forms/d/e/1FAIpQLScyeRsF2GJjYdf9bUyeDjt8_9id-gvqiBU1SHR3ioDGe5eF3g/viewform?usp=header">this form</a> with details. 
+
```
```suggestion
```

sstonehill12 commented:
/merge
